pkg/semtech-loramac/contrib/semtech_loramac.c

```diff
@@ -708,3 +715,93 @@ uint8_t semtech_loramac_recv(semtech_loramac_t *mac)
+
+    /* Store EUIs, Keys and device address */
+    pos += eeprom_write(pos, mac->deveui, LORAMAC_DEVEUI_LEN);
+    pos += eeprom_write(pos, mac->appeui, LORAMAC_APPEUI_LEN);
+    pos += eeprom_write(pos, mac->appkey, LORAMAC_APPKEY_LEN);
+    pos += eeprom_write(pos, mac->appskey, LORAMAC_APPSKEY_LEN);
+    pos += eeprom_write(pos, mac->nwkskey, LORAMAC_NWKSKEY_LEN);
+    pos += eeprom_write(pos, mac->devaddr, LORAMAC_DEVADDR_LEN);
+
+    /* uplink counter, mainly used for ABP activation */
+    uint8_t counter[4];
+    mutex_lock(&mac->lock);
+    MibRequestConfirm_t mibReq;
+    mibReq.Type = MIB_UPLINK_COUNTER;
+    LoRaMacMibGetRequestConfirm(&mibReq);
+    *(uint32_t *)counter = mibReq.Param.UpLinkCounter;
+    mutex_unlock(&mac->lock);
+    pos += eeprom_write(pos, counter, 4);
+}
+
+void _set_uplink_counter(semtech_loramac_t *mac, uint8_t *counter)
+{
+    DEBUG("[semtech-loramac] uplink counter: %lu\n", *(uint32_t *)counter);
+
+    mutex_lock(&mac->lock);
+    MibRequestConfirm_t mibReq;
+    mibReq.Type = MIB_UPLINK_COUNTER;
+    mibReq.Param.UpLinkCounter = *(uint32_t *)counter;
+    LoRaMacMibSetRequestConfirm(&mibReq);
+    mutex_unlock(&mac->lock);
+}
+
+int semtech_loramac_read_config(semtech_loramac_t *mac)
+{
+    DEBUG("[semtech-loramac] reading configuration from EEPROM\n");
+
+    uint8_t magic[SEMTECH_LORAMAC_EEPROM_MAGIC_LEN];
+    size_t pos = SEMTECH_LORAMAC_EEPROM_START;
+    pos += eeprom_read(pos, magic, SEMTECH_LORAMAC_EEPROM_MAGIC_LEN);
+    if (strcmp((char*)magic, "RIOT") != 0) {
+        DEBUG("[semtech-loramac] no configuration present on EEPROM\n");
+        return -1;
+    }
+
+    /* LoRaWAN EUIs, Keys and device address */
+    pos += eeprom_read(pos, mac->deveui, LORAMAC_DEVEUI_LEN);
+    pos += eeprom_read(pos, mac->appeui, LORAMAC_APPEUI_LEN);
+    pos += eeprom_read(pos, mac->appkey, LORAMAC_APPKEY_LEN);
+    pos += eeprom_read(pos, mac->appskey, LORAMAC_APPSKEY_LEN);
+    pos += eeprom_read(pos, mac->nwkskey, LORAMAC_NWKSKEY_LEN);
+    pos += eeprom_read(pos, mac->devaddr, LORAMAC_DEVADDR_LEN);
+
+    /* uplink counter, mainly used for ABP activation */
+    uint8_t counter[4] = { 0 };
+    pos += eeprom_read(pos, counter, 4);
+    _set_uplink_counter(mac, counter);
+
+    return 0;
+}
+
+void semtech_loramac_erase_config(void)
+{
+    DEBUG("[semtech-loramac] erasing configuration on EEPROM\n");
+
+    uint8_t magic[SEMTECH_LORAMAC_EEPROM_MAGIC_LEN];
+    size_t pos = SEMTECH_LORAMAC_EEPROM_START;
+    eeprom_read(pos, magic, SEMTECH_LORAMAC_EEPROM_MAGIC_LEN);
+    if (strcmp((char*)magic, "RIOT") != 0) {
+        DEBUG("[semtech-loramac] no configuration present on EEPROM\n");
+        return;
+    }
+
+    size_t end = (pos + SEMTECH_LORAMAC_EEPROM_MAGIC_LEN +
+                  LORAMAC_DEVEUI_LEN + LORAMAC_APPEUI_LEN +
+                  LORAMAC_APPKEY_LEN + LORAMAC_APPSKEY_LEN +
+                  LORAMAC_NWKSKEY_LEN + LORAMAC_DEVADDR_LEN + 4);
```

Comment:
what does this `4` in the end represents?

Reply:
These are the 4 bytes used for the uplink counter (uint32_t)

Reply:
maybe define it as a LORAMAC_UPCOUNTER_LEN?